pkg/microservice/aslan/core/system/service/registry.go

```diff
@@ -199,44 +216,293 @@ func ListAllRepos(log *zap.SugaredLogger) ([]*RepoInfo, error) {
+					canWrite.m.Lock()
+					if *canWrite.status == 1 {
+						errChan <- fmt.Errorf("getting image list, please try again later")
+					}
+					canWrite.m.Unlock()
+					// continue to get next repo to update image tags db
+					continue
+				}
+			} else {
+				logger.Errorf("find image[%s] tags from db failed, the error is: %v", repo.Name, err)
+			}
+		} else {
+			for _, imageTag := range imageTags.ImageTags {
+				dbTags[imageTag.TagName] = imageTag
+			}
+		}
+		logger.Infof("get image[%s] %d tags from db, get image[%s] %d tags from registry", repo.Name, len(dbTags), repo.Name, len(repo.Tags))
+
+		tags := ImageListGetter(repo, registryInfo, dbTags, regService, logger)
+		// update all imageTags to db,
+		go insertNewTags2DB(tags, registryInfo, key, logger)
+
+		images = append(images, tags...)
+	}
+
+	canWrite.m.Lock()
+	if *canWrite.status == 1 {
+		result <- images
+	}
+	canWrite.m.Unlock()
+}
+
+type newImages struct {
+	resp []*RepoImgResp
+	m    *sync.Mutex
+}
+
+func ImageListGetter(repo *registry.Repo, registryInfo *commonmodels.RegistryNamespace, dbTags map[string]*models.ImageTag, regService registry.Service, logger *zap.SugaredLogger) []*RepoImgResp {
+	resultChan := make(chan *RepoImgResp, 10)
+	defer close(resultChan)
+	jobChan := make(chan *imageReqJob, len(repo.Tags))
+	images := &newImages{
+		resp: make([]*RepoImgResp, 0),
+		m:    &sync.Mutex{},
+	}
+
+	jobCount := 0
+	for _, tag := range repo.Tags {
+		if t, ok := dbTags[tag]; ok && t.Digest != "" && t.Created != "" {
+			resp := &RepoImgResp{
+				Host:    util.TrimURLScheme(registryInfo.RegAddr),
+				Owner:   repo.Namespace,
+				Name:    repo.Name,
+				Tag:     t.TagName,
+				Created: t.Created,
+				Digest:  t.Digest,
+			}
+			images.resp = append(images.resp, resp)
+			continue
+		}
+
+		jobCount++
+		jobChan <- &imageReqJob{
+			regService: regService,
+			endPoint: registry.Endpoint{
+				Addr:      registryInfo.RegAddr,
+				Ak:        registryInfo.AccessKey,
+				Sk:        registryInfo.SecretKey,
+				Namespace: registryInfo.Namespace,
+				Region:    registryInfo.Region,
+			},
+			registryInfo: registryInfo,
+			repoImages:   repo,
+			tag:          tag,
+		}
+	}
+	// close jobChan to notify all goroutines to exit when they finish all jobs
+	close(jobChan)
+
+	if jobCount > 0 {
+		// create gcount goroutines to get tags from registry
+		gcount := 50
+		if jobCount < 50 {
+			gcount = jobCount
+		}
+		wg := &sync.WaitGroup{}
+		for i := 0; i < gcount; i++ {
+			wg.Add(1)
+			go getNewTagsFromReg(jobChan, wg, images, logger)
+		}
+
+		wg.Wait()
+	}
+
+	return images.resp
+}
+
+func getNewTagsFromReg(job chan *imageReqJob, wg *sync.WaitGroup, images *newImages, logger *zap.SugaredLogger) {
+	defer wg.Done()
+
+	for data := range job {
+		option := registry.GetRepoImageDetailOption{
+			Image:    data.repoImages.Name,
+			Tag:      data.tag,
+			Endpoint: data.endPoint,
+		}
+		details, err := data.regService.GetImageInfo(option, logger)
+		if err != nil {
```

Comment:
suggest to log an warn here.

Reply:
i think that write one error log is enough if it can not retry.